### DISCUSSION TITLE
Get the raw format from serde::format_description!

### DISCUSSION THREAD

**Discussion by Nohac** [time-rs/time#515] [Question] [answered]
Hi. 

When doing the following:
```rust
time::serde::format_description!(
    date_time_format,
    PrimitiveDateTime,
    "[year]-[month]-[day]T[hour]:[minute]:[second].[subsecond digits:3]"
);
```
Is it possible to access to format to be used for turning a date back into string?
What I'm currently doing is to define the format twice, which causes duplicate code:

```rust
const DATE_TIME_FOMRAT: &[FormatItem<'_>] = time::macros::format_description!(
    "[year]-[month]-[day]T[hour]:[minute]:[second].[subsecond digits:3]"
);

...

let str_ts = timestamp.format(&format).unwrap().to_string();

```
Instead, I would prefer to do something like this:
```rust
let str_ts = timestamp.format(&date_time_format::format).unwrap().to_string();
```

Also, slightly unrelated question, but is it possible to define "optional" time components? In the example above, I want the `.[subsecond digits:3]` to be optional.

**Comment by jhpratt** [comment, marked as answer]
While the format description string is duplicated, the actual duplication in generated code is insignificant, as `const`s are effectively values that are inlined wherever they're used. With regard to exposing the value, this is doable. It's already present, just scoped to the generated module. I'll think about it some.

> Also, slightly unrelated question, but is it possible to define "optional" time components? In the example above, I want the `.[subsecond digits:3]` to be optional.

Not currently. It is tracked in #428, where there is no update.

**Reply by Nohac** [reply]
Thanks.
I'm not really concerned about the potential generated code duplication, I'm more concerned about having to keep multiple formats up to date that's supposed to be the same, so I'd really appreciate if you could make that value accessible, or as an alternative, allow passing a format to `serde::format_description!` directly:
```rust
const DATE_TIME_FOMRAT: &[FormatItem<'_>] = time::macros::format_description!(
    "[year]-[month]-[day]T[hour]:[minute]:[second].[subsecond digits:3]"
);

time::serde::format_description!(
    date_time_format,
    PrimitiveDateTime,
    DATE_TIME_FOMRAT
);
```

**Reply by jhpratt** [reply]
I figured that's what you were referring to, but wasn't certain. A fundamental limitation of macros is that they are expanded before resolution, so passing a `const` as a parameter is impossible. If anything changes, it will be the format description being made available.

**Reply by jhpratt** [reply]
_Actually_, thinking about this again, I could reasonably detect if it's a string or not. If it's not a string, I could interpret all remaining tokens as a path and use that instead of parsing and emitting a new format description.

It's a wild thought, but I'm fairly certain it's doable! Probably the cleanest option in my opinion.

**Reply by jhpratt** [reply]
Hmm…hygiene is an issue now. I think there's a workaround but it's a big hacky. I'll give it a shot in the coming days, most likely. Other than hygiene it's complete.

**Reply by jhpratt** [reply]
Turns out it wasn't hygiene. Regardless, I've just pushed up the change! Try cloning the main branch and play around with it. I added a simple test case to the crate's test suite, so it should work without issue. The only requirement is that the path be `const _: &[FormatItem<'_>]`.
